### PR TITLE
Maintain version order from xml

### PR DESCRIPTION
**Pull request by FinlayRJW**
## Before this PR
The order that versions appeared in as part of a suggestion would be determined by the `sort` order of the strings of the versions

## After this PR
We now use the `PrioritizedLookupElement` to force the order to be in that of the xml e.g. descending order
==COMMIT_MSG==
Maintain version order from xml
==COMMIT_MSG==

## Possible downsides?
If versions are added from a different contributor they will not automatically gain this ordering as it is added at the contributor level

### CODE REVIEW

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/RepositoryExplorer.java [hunk 1]
```diff
@@ -102,8 +104,8 @@ private Set<GroupPartOrPackageName> fetchFoldersFromContent(String contents) {
         return folders;
     }
 
-    private Set<DependencyVersion> parseVersionsFromContent(String content) {
-        Set<DependencyVersion> versions = new HashSet<>();
+    private List<DependencyVersion> parseVersionsFromContent(String content) {
```
Comment (felixdesouza) [inline gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/RepositoryExplorer.java]: I think you should keep this as a set, because now you're no longer checking for duplicates. Where you're using it below, is just converting it to some sort of iterable anyway, so this change doesn't do anything for you there.

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java [hunk 2]
```diff
@@ -67,7 +70,14 @@ public void addCompletions(
                                                         .withLookupString("latest"),
                                                 Double.MAX_VALUE)
                                         : LookupElementBuilder.create(version))
-                                .forEach(resultSet::addElement);
+                                .toList();
+
+                        List<LookupElement> prioritizedVersions = IntStream.range(0, versions.size())
+                                .mapToObj(i ->
+                                        PrioritizedLookupElement.withPriority(versions.get(i), i - versions.size()))
+                                .toList();
```
Comment (felixdesouza) [inline gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java]: will this not override the priority set for the `latest` elements?

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/RepositoryExplorer.java (hunk 1 above)
Comment (felixdesouza) [reply gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/RepositoryExplorer.java]: if you're looking for order preserving set, you could use `LinkedHashSet` instead

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java (hunk 2 above)
Comment (felixdesouza) [reply gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java]: if you used `StreamEx` from line 64, you can actually just zip with an infinite stream that counts down. 

Doesn't really matter if it starts at 0 since you want relative order

e.g. 
```java
StreamEx(...)
  .flatMap(...)
  .zipWith(IntStream.iterate(0, i -> i - 1)
  .map((version, priority) -> getLookupElement(version, priority)
  .forEach(resultSet::addElement);
```

**Review by felixdesouza — COMMENTED** [review]
you should add some tests to verify the behaviour!

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java (hunk 2 above)
Comment (FinlayRJW) [reply gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java]: switched to streamex

**Comment by FinlayRJW** [comment]
> you should add some tests to verify the behaviour!

I have added a test to confirm that the versions order is correct 😄

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java [hunk 3]
```diff
@@ -58,17 +61,22 @@ public void addCompletions(
...
                     }
+
+                    private LookupElement getLookupElement(DependencyVersion version, Integer priority) {
+                        return version.isLatest()
+                                ? PrioritizedLookupElement.withPriority(
+                                        LookupElementBuilder.create(version)
+                                                .withTypeText("Latest", true)
+                                                .withLookupString("latest"),
+                                        Double.MAX_VALUE)
+                                : PrioritizedLookupElement.withPriority(LookupElementBuilder.create(version), priority);
```
Comment (CRogers) [inline gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java:78]: This is making it so that if the repos are:

* jar
* dist


then all jar completions will appear before dist ones?

If so I'm not sure this is super necessary. It could be from either repo.

Comment (CRogers) [reply gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java:78]: Ah, it forces the order to be descending too, makes sense.

File: gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java [hunk 4]
```diff
@@ -58,17 +61,22 @@ public void addCompletions(
 
                         Project project = parameters.getOriginalFile().getProject();
 
-                        RepositoryLoader.loadRepositories(project).stream()
+                        StreamEx.of(RepositoryLoader.loadRepositories(project))
                                 .flatMap(url -> repositoryExplorer.getVersions(group, dependencyPackage, url).stream())
```
Comment (CRogers) [inline gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java:65]: Might be worth parallelising this somehow in another PR.

**Review by CRogers — APPROVED** [review]
